arbnode/dataposter/data_poster.go

```diff
@@ -139,18 +146,60 @@ func NewDataPoster(ctx context.Context, opts *DataPosterOpts) (*DataPoster, erro
 	default:
 		queue = slice.NewStorage(func() storage.EncoderDecoderInterface { return &storage.EncoderDecoder{} })
 	}
-	return &DataPoster{
-		headerReader:      opts.HeaderReader,
-		client:            opts.HeaderReader.Client(),
-		sender:            opts.Auth.From,
-		signer:            opts.Auth.Signer,
+	dp := &DataPoster{
+		headerReader: opts.HeaderReader,
+		client:       opts.HeaderReader.Client(),
+		sender:       opts.Auth.From,
+		signer: func(_ context.Context, addr common.Address, tx *types.Transaction) (*types.Transaction, error) {
+			return opts.Auth.Signer(addr, tx)
+		},
 		config:            opts.Config,
 		replacementTimes:  replacementTimes,
 		metadataRetriever: opts.MetadataRetriever,
 		queue:             queue,
 		redisLock:         opts.RedisLock,
 		errorCount:        make(map[uint64]int),
-	}, nil
+	}
+	if cfg.ExternalSigner.Enabled {
+		signer, sender, err := externalSigner(ctx, &cfg.ExternalSigner)
+		if err != nil {
+			return nil, err
+		}
+		dp.signer, dp.sender = signer, sender
+	}
+	return dp, nil
+}
+
+// externalSigner returns signer function and ethereum address of the signer.
+// Returns an error if address isn't specified or if it can't connect to the
+// signer RPC server.
+func externalSigner(ctx context.Context, opts *ExternalSignerCfg) (signerFn, common.Address, error) {
+	if opts.Address == "" {
+		return nil, common.Address{}, errors.New("external signer (From) address specified")
+	}
+	sender := common.HexToAddress(opts.Address)
+	client, err := rpc.DialContext(ctx, opts.URL)
+	if err != nil {
+		return nil, common.Address{}, fmt.Errorf("error connecting external signer: %w", err)
+	}
+	return func(ctx context.Context, addr common.Address, tx *types.Transaction) (*types.Transaction, error) {
+		// According to the "eth_signTransaction" API definition, this shoul be
+		// RLP encoded transaction object.
+		// https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_signtransaction
+		var rlpEncTxStr string
+		if err := client.CallContext(ctx, &rlpEncTxStr, "eth_signTransaction", tx); err != nil {
+			return nil, fmt.Errorf("signing transaction: %w", err)
+		}
+		data, err := hexutil.Decode(rlpEncTxStr)
+		if err != nil {
+			return nil, fmt.Errorf("error decoding hex: %w", err)
+		}
+		var signedTx types.Transaction
+		if err := rlp.DecodeBytes(data, &signedTx); err != nil {
+			return nil, fmt.Errorf("error decoding signed transaction: %w", err)
+		}
+		return &signedTx, nil
```

Comment:
Good point! Added the check, although doing the lazy initialization, otherwise passing the signer (or chain ID) all the way from nitro.go would unnecessarily pollute call chain.